### PR TITLE
Properly handle fatal template diagnostics for Angular CLI

### LINKED ISSUE #27045: `FatalDiagnosticError` is being thrown by the compiler when removing signal input or `Child` component is not exported
### Command

build, serve

### Is this a regression?

- [ ] Yes, this behavior used to work in the previous version

### The previous version in which this bug was not present was

17.1

### Description

The CLI throws the following error when running the provided repro. 

```
[✘ [ERROR] Cannot read properties of null (reading 'errors') [plugin angular-compiler]

    node_modules/@angular-devkit/build-angular/src/tools/esbuild/angular/compiler-plugin.js:199:32:
      199 │                 if (diagnostics.errors?.length) {
          ╵                                 ^

    at /Users/kyro/Developer/web/angular-projects/next/node_modules/@angular-devkit/build-angular/src/tools/esbuild/angular/compiler-plugin.js:199:33
    at async /Users/kyro/Developer/web/angular-projects/next/node_modules/esbuild/lib/main.js:1350:22
    at async Promise.all (index 0)
    at async requestCallbacks.on-start (/Users/kyro/Developer/web/angular-projects/next/node_modules/esbuild/lib/main.js:1348:5)
    at async handleRequest (/Users/kyro/Developer/web/angular-projects/next/node_modules/esbuild/lib/main.js:732:11)

  This error came from the "onStart" callback registered here:

    node_modules/@angular-devkit/build-angular/src/tools/esbuild/angular/compiler-plugin.js:81:18:
      81 │             build.onStart(async () => {
         ╵                   ~~~~~~~

    at setup (/Users/kyro/Developer/web/angular-projects/next/node_modules/@angular-devkit/build-angular/src/tools/esbuild/angular/compiler-plugin.js:81:19)
    at async handlePlugins 

```
### Minimal Reproduction

Repro : https://github.com/jeanmeche/next 

The issue doesn't not happen when either 

* The `Child` component is exported
* Signal input is removed. 

While there is maybe an issue on the signal implementation part, the CLI probably shouldn't throw the way it does. 

### Your Environment

```text
_                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/
    

Angular CLI: 17.1.2
Node: 18.17.0
Package Manager: npm 9.6.7
OS: darwin x64

Angular: 17.1.2
... animations, cli, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, platform-server
... router, ssr

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1701.2
@angular-devkit/build-angular   17.1.2
@angular-devkit/core            17.1.2
@angular-devkit/schematics      17.1.2
@schematics/angular             17.1.2
rxjs                            7.8.1
typescript                      5.3.3
zone.js                         0.14.3
```

### LINKED ISSUE #53678: Error: Module build failed: [object Object]
### Which @angular/* package(s) are the source of the bug?

core

### Is this a regression?

Yes

### Description

When passing a number in a component:

```typescript
import "zone.js";
import { bootstrapApplication } from "@angular/platform-browser";

import { Component, Input } from "@angular/core";

@Component({
	selector: "display-size",
	standalone: true,
	template: `<p>Test</p>`,
})
class DisplaySizeComponent {
	@Input() bytes!: number;
}

@Component({
	selector: "app-root",
	standalone: true,
	imports: [DisplaySizeComponent],
	template: `
		<display-size [bytes]="123" />
	`,
})
export class AppComponent {}

bootstrapApplication(AppComponent);
```

It seems to throw an `[object Object]` error

### Please provide a link to a minimal reproduction of the bug

https://stackblitz.com/github/crutchcorn/angular-object-object-bug?file=src%2Fmain.ts&template=node

### Please provide the exception or error you saw

```true
Build at: 2023-12-21T17:46:12.035Z - Hash: cc1a53dd72abbe97 - Time: 3228ms

./src/main.ts - Error: Module build failed (from ./node_modules/@ngtools/webpack/src/ivy/index.js):
Error: [object Object]
```

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: 17.0.7
Node: 18.18.0
Package Manager: npm 9.4.2
OS: linux x64

Angular: 17.0.8
... animations, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1700.7
@angular-devkit/build-angular   17.0.7
@angular-devkit/core            17.0.7
@angular-devkit/schematics      17.0.7
@angular/cli                    17.0.7
@schematics/angular             17.0.7
rxjs                            7.8.1
typescript                      5.2.2
zone.js                         0.14.2
```

### Anything else?

I sincerely apologize if I'm doing something wrong. I'm working with a lot of frameworks at the moment so it's possible I'm using a wrong syntax or something. I promise I double and triple checked

### PR DESCRIPTION
**Pull request by devversion**
See individual commits.

Fixes #53678. Fixes https://github.com/angular/angular-cli/issues/27045

### CODE REVIEW

File: packages/compiler-cli/src/ngtsc/core/src/compiler.ts [hunk 1]
```diff
@@ -444,21 +461,22 @@ export class NgCompiler {
...
-    } catch (e) {
-      // Type check code may throw fatal diagnostic errors if e.g. the type check
-      // block cannot be generated. Gracefully return the associated diagnostic.
-      // Note: If a fatal diagnostic is raised, do not repeat the same diagnostics
-      // by running the extended template checking code, which will attempt to
-      // generate the same TCB.
-      if (e instanceof FatalDiagnosticError) {
-        diagnostics.push(e.toDiagnostic());
+    } catch (err: unknown) {
+      if (!(err instanceof FatalDiagnosticError)) {
```
Comment (alan-agius4) [inline packages/compiler-cli/src/ngtsc/core/src/compiler.ts:476]: You could also use `isFatalDiagnosticError`

**Review by alan-agius4 — APPROVED** [review]
LGTM

File: packages/compiler-cli/src/ngtsc/core/src/compiler.ts (hunk 1 above)
Comment (devversion) [reply packages/compiler-cli/src/ngtsc/core/src/compiler.ts]: Done! like it

File: packages/compiler-cli/test/ngtsc/template_typecheck_spec.ts [hunk 2]
```diff
@@ -69,6 +69,34 @@ export declare class AnimationEvent {
       expect(getSourceCodeForDiagnostic(diags[0])).toBe('does_not_exist');
     });
 
+    it('should not fail with a runtime error when generating TCB', () => {
```
[the remark this answers is not in the document]
Comment (devversion) [reply packages/compiler-cli/test/ngtsc/template_typecheck_spec.ts:72]: It's: Type check block. It's a common term used across the compiler.

**Review by crisbeto — APPROVED** [review]

**Comment by jessicajaniuk** [comment]
Caretaker note: failures are unrelated. Safe to merge.

**Comment by jessicajaniuk** [comment]
This PR was merged into the repository by commit a08863e7e029a95d4dfb425ad45f438a1d9c1f3d.

**Pull request closed without merging**